### PR TITLE
Implement typedesc for mapping binding pattern with anonymous record type

### LINKED ISSUE #39633: [Bug]: Mapping binding pattern fails for anonymous record type
### Description

Consider the following code
```ballerina
import ballerina/io;

type Representation record {
    string __typename;
};

public function main() returns error? {
    Representation representation = {__typename: "Product", "variation": {id: "10"}};
    record {
        string id?;
        string sku?;
        string package?;
        record {string id;} variation?;
    } {id, sku, package, variation} = check representation.ensureType(); // this crashes

    io:println(id, variation?.id);
}
```
The above code crashes with following error
```
[2023-02-16 12:59:49,370] SEVERE {b7a.log.crash} - $typedesc$$anonType$_0 
java.lang.NoClassDefFoundError: $typedesc$$anonType$_0
	at test.main(test.bal:0)
	at $_init.$lambda$main$(.)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:324)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:291)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassNotFoundException: $typedesc$$anonType$_0
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 6 more
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.4.0 RC2, 2201.3.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #39633 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal [hunk 1]
```diff
@@ -743,6 +743,20 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
...
+function recordVariableWithAnonymousType() returns [string, boolean]|error {
+    record {|
+        string name;
+        boolean married;
+    |} {name, married} = check getPersonRecord().ensureType();
+    return [name, married];
+}
+
+function testRecordVariableWithAnonymousType() {
+    [string, boolean]|error [name, married] = recordVariableWithAnonymousType();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:755]: This compiling is wrong, isn't it? See the example in https://github.com/ballerina-platform/ballerina-lang/issues/39690. 

Let's rewrite this test to not rely on the bug.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal [hunk 2]
```diff
@@ -743,6 +743,20 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyRecord> {x: [1, 2], y: "s1"}, b1);
 }
 
+function recordVariableWithAnonymousType() returns [string, boolean]|error {
+    record {|
+        string name;
+        boolean married;
+    |} {name, married} = check getPersonRecord().ensureType();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:750]: Do we have tests for tuples and errors?

Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:750]: Shall we add a few more tests with nested patterns?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:750]: We currently do not allow a record literal of that type with the binding pattern[example 1], so we use the 'typedesc' of the binding pattern when used with the 'ensure' type.

```ballerina
function recordVariableWithAnonymousType() returns [string, boolean] {
    record {|
        string name;
        boolean married;
    |} {name, married} = {name: "Jack", married: true};
    return [name, married];
}
```

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:750]: Are there any other scenarios where we might need to use the 'typedesc' of a binding pattern?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:750]: I don't get the relevance? I meant tests for list and error binding patterns with tuples (and maybe arrays) and errors that have anonymous records/mapping binding patterns as members?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: I have added a test case for a nested scenario. Could you please confirm whether this is the type of scenario you intended?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Could you please provide an example for this?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]:
```ballerina
    [int, record {|
        string name;
        boolean married;
    |}] [id, {name, married}] = check a.ensureType();
```

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: In https://github.com/ballerina-platform/ballerina-lang/pull/39688/commits/5a0b98caf6c71532c6d80eb619caac6a39588f04#diff-ee5929c4cfdb578fc7e234b7577be5fa4aa66a2687e2b4ffa65fe35cfc62b889R775 we can use another mapping binding pattern for details.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal [hunk 3]
```diff
@@ -743,6 +743,55 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
...
+function recordVariableWithAnonymousType() returns [string, boolean] {
+    do {
+        record {|
+            string name;
+            boolean married;
+        |} {name, married} = check getPersonRecord().ensureType();
+        return [name, married];
+    } on fail {
+        return ["name", false];
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:755]: Why do we need to do all this? Can't we use checkpanic and simplify?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal (hunk 2 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Why do we need to generate a typedesc for?
record {|
        string name;
        boolean married;
    |} in LHS

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: The typedesc used as the argument for `ensureType` is 

```
[int, record {|
        string name;
        boolean married;
    |}]
```

which has the record as the member.

Why shouldn't we generate anyway, when there is a record typedesc in the code?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Yes, in this scenario we generate the 'typedesc' and create the tuple variable. However, we encounter this problem when visiting the record variable.
We have these types of tests in https://github.com/ballerina-platform/ballerina-lang/blob/master/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: If the type node wasn't visited for "tuple variables", we'd have a similar issue, right? Thus, my concern about having a test for tuples/errors too.

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: In the current implementation we visited the typenode of tuple variables https://github.com/ballerina-platform/ballerina-lang/blob/3eda82f07b8e47b8ade714deae35dc2574504f1e/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L1522

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Yes, my point is, if we have tests for them?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Yes. We have tests in  https://github.com/ballerina-platform/ballerina-lang/blob/master/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Here, typeDesc of record type is used to create TupleValueImp. So do we need to add a test with ensureType?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: add test for this.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal [hunk 4]
```diff
@@ -743,6 +743,47 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyRecord> {x: [1, 2], y: "s1"}, b1);
 }
 
+function recordVariableWithAnonymousType() returns [string, boolean] {
+    record {|
+        string name;
+        boolean married;
+    |} {name, married} = checkpanic getPersonRecord().ensureType();
+    return [name, married];
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:752]: We don't need separate functions. right? Can be within the test functions itself?

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal]: Updated

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**